driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java

```diff
@@ -50,30 +66,89 @@
-        private Simple(Host host) {
+    static class DefaultConvictionPolicy extends ConvictionPolicy {
+        private final Host host;
+        private final ReconnectionPolicy reconnectionPolicy;
+        private final AtomicInteger openConnections = new AtomicInteger();
+
+        private volatile long nextReconnectionTime = Long.MIN_VALUE;
+        private ReconnectionPolicy.ReconnectionSchedule reconnectionSchedule;
+
+        private DefaultConvictionPolicy(Host host, ReconnectionPolicy reconnectionPolicy) {
+            this.host = host;
+            this.reconnectionPolicy = reconnectionPolicy;
+        }
+
+        @Override
+        void signalConnectionCreated() {
+            int newCount = openConnections.incrementAndGet();
+            Host.statesLogger.debug("[{}] new connection created, total = {}", host, newCount);
+            resetReconnectionTime();
+        }
+
+        @Override
+        void signalConnectionClosed() {
+            if (host.state == Host.State.DOWN)
+                return;
+            int remaining = openConnections.decrementAndGet();
+            assert remaining >= 0;
+            Host.statesLogger.debug("[{}] connection closed, remaining = {}", host, remaining);
         }
 
         @Override
-        public boolean addFailure(ConnectionException exception) {
-            return true;
+        boolean signalConnectionFailure(ConnectionException exception, boolean wasFullyInitialized) {
+            if (host.state == Host.State.DOWN)
+                return false;
+
+            updateReconnectionTime();
+            if (wasFullyInitialized) {
+                int remaining = openConnections.decrementAndGet();
+                assert remaining >= 0;
+                Host.statesLogger.debug("[{}] remaining connections = {}", host, remaining);
+                return remaining == 0;
+            } else {
+                return false;
+            }
+        }
+
+        private synchronized void updateReconnectionTime() {
+            long now = System.currentTimeMillis();
+            if (nextReconnectionTime > now)
+                // Someone else updated the time before us
+                return;
+
+            if (reconnectionSchedule == null)
+                reconnectionSchedule = reconnectionPolicy.newSchedule();
+
+            long nextDelayMs = reconnectionSchedule.nextDelayMs();
+            Host.statesLogger.debug("[{}] preventing new connections for the next {} ms", host, nextDelayMs);
+            nextReconnectionTime = now + nextDelayMs;
         }
 
-        public boolean addFailureFromExternalDetector() {
-            return true;
+        private synchronized void resetReconnectionTime() {
+            reconnectionSchedule = null;
+            nextReconnectionTime = Long.MIN_VALUE;
         }
 
         @Override
-        public void reset() {}
+        boolean canReconnectNow() {
+            return System.currentTimeMillis() >= nextReconnectionTime;
```

Comment:
We _could_ extract `updateReconnectionTime`, `resetReconnectionTime` and `canReconnectNow` somewhere else, but I'm not convinced. IMO the real issue is that `ConvictionPolicy` should probably be renamed. But it's about collecting failures and deciding when you can reconnect based on past failures, so it doesn't shock me that both responsibilities go in the same class.

`ReconnectionPolicy` only manages the reconnection schedule, not what we do with them. Plus changing it would break binary compatibility.

I don't think we should expose `ConvictionPolicy`. In fact I'm tempted to get rid of the abstract class and factory. As noted by Sylvain in the comments, "maybe just exposing say a threshold of error before convicting a node is enough".

> wouldn't System.nanoTime() be a better choice

Agreed.